Debian 11 Source DKMS Build Fails Due to 'alien' and 'dh'

### System information

| Type | Version/Name |
| ----- | -------------- |
| Distribution Name | Debian |
| Distribution Version | 11 Bullseye/Testing (Soft Freeze) |
| Linux Kernel |  5.10.0-3-amd64 |
| Architecture | amd64 |
| ZFS Version | 2.0.3 |
| SPL Version | Unknown |

### Describe the problem you're observing

On a fresh installation of Debian 11 Testing (currently in a Soft Freeze), it is impossible to build the required *.deb packages due to an error with `alien` (which actually shows an error from `dh`). Apparently the build process first creates RPMs (regardless of the distribution) and then tries (and fails) to convert them to other formats.

On a somewhat related note, I was able to build OpenZFS 2.0.0 in the past on Debian 10 Buster/Stable without any issues.

### Describe how to reproduce the problem

I am following the [Debian and Ubuntu DKMS installation guide](https://openzfs.github.io/openzfs-docs/Developer%20Resources/Custom%20Packages.html#debian-and-ubuntu). Download OpenZFS 2.0.3 on Debian 11 and build with these settings:

```
$ ./configure --enable-systemd
$ make -j $(nproc) deb-utils deb-dkms
```

### Include any warning/errors/backtraces from the system logs

Here is the actual error reported by the `make` command:

```
make[1]: Entering directory '/root/zfs-2.0.3/zfs-dkms-2.0.3'
dh
dh: error: specify a sequence to run
make[1]: *** [debian/rules:7: binary] Error 25
make[1]: Leaving directory '/root/zfs-2.0.3/zfs-dkms-2.0.3'
make: *** [Makefile:1695: deb-dkms] Error 1
```

This error also occurs when using the `alien` command manually. Is [this the official alien project](https://github.com/mildred/alien)? It has been untouched since 2011.

```
$ alien zfs-2.0.3-1.x86_64.rpm
Warning: Skipping conversion of scripts in package zfs: postinst postrm prerm
Warning: Use the --scripts parameter to include the scripts.
Package build failed. Here's the log:
dh
dh: error: specify a sequence to run
make: *** [debian/rules:7: binary] Error 25
```

As an alternative, I tried using [fpm](https://github.com/jordansissel/fpm) but it failed to properly convert the packages. Various errors existed with the *.deb packages it created.

```
$ for i in $(ls -1 | grep -P '\.rpm' | grep -v -P '\.src'); do fpm -s rpm -t deb $i; done
$ sudo dpkg -i ./*.deb
dpkg: error processing archive ./libnvpair3_2.0.3-1_amd64.deb (--install):
 parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'libnvpair3':
 'Depends' field, invalid package name '/sbin/ldconfig': must start with an alphanumeric character
dpkg: error processing archive ./libuutil3_2.0.3-1_amd64.deb (--install):
 parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'libuutil3':
 'Depends' field, invalid package name '/sbin/ldconfig': must start with an alphanumeric character
dpkg: error processing archive ./libzfs4_2.0.3-1_amd64.deb (--install):
 parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'libzfs4':
 'Depends' field, invalid package name '/sbin/ldconfig': must start with an alphanumeric character
dpkg: warning: parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'libzfs4-devel':
 'Depends' field, reference to 'rpmlib':
 implicit exact match on version number, suggest using '=' instead
dpkg: error processing archive ./libzfs4-devel_2.0.3-1_amd64.deb (--install):
 parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'libzfs4-devel':
 'Depends' field, reference to 'rpmlib': version 'CompressedFileNames': version number does not start with digit
dpkg: error processing archive ./libzpool4_2.0.3-1_amd64.deb (--install):
 parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'libzpool4':
 'Depends' field, invalid package name '/sbin/ldconfig': must start with an alphanumeric character
dpkg: warning: parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'python3-pyzfs':
 'Depends' field, reference to 'rpmlib':
 implicit exact match on version number, suggest using '=' instead
dpkg: error processing archive ./python3-pyzfs_2.0.3-1_all.deb (--install):
 parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'python3-pyzfs':
 'Depends' field, reference to 'rpmlib': version 'CompressedFileNames': version number does not start with digit
dpkg: error processing archive ./zfs-dkms_2.0.3-1_all.deb (--install):
 parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'zfs-dkms':
 'Depends' field, invalid package name '/bin/sh': must start with an alphanumeric character
dpkg: error processing archive ./zfs-dracut_2.0.3-1_all.deb (--install):
 parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'zfs-dracut':
 'Depends' field, invalid package name '/bin/sh': must start with an alphanumeric character
dpkg: error processing archive ./zfs-initramfs_2.0.3-1_amd64.deb (--install):
 parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'zfs-initramfs':
 'Depends' field, invalid package name '/bin/sh': must start with an alphanumeric character
dpkg: warning: parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'zfs-test':
 'Depends' field, reference to 'rpmlib':
 implicit exact match on version number, suggest using '=' instead
dpkg: error processing archive ./zfs-test_2.0.3-1_amd64.deb (--install):
 parsing file '/var/lib/dpkg/tmp.ci/control' near line 7 package 'zfs-test':
 'Depends' field, reference to 'rpmlib': version 'CompressedFileNames': version number does not start with digit
Errors were encountered while processing:
 ./libnvpair3_2.0.3-1_amd64.deb
 ./libuutil3_2.0.3-1_amd64.deb
 ./libzfs4_2.0.3-1_amd64.deb
 ./libzfs4-devel_2.0.3-1_amd64.deb
 ./libzpool4_2.0.3-1_amd64.deb
 ./python3-pyzfs_2.0.3-1_all.deb
 ./zfs-dkms_2.0.3-1_all.deb
 ./zfs-dracut_2.0.3-1_all.deb
 ./zfs-initramfs_2.0.3-1_amd64.deb
 ./zfs-test_2.0.3-1_amd64.deb
```

I am experiencing this problem in Debian bullseye/testing, as well.  I get this right after `make` attempts `dh` +  `alien`:

```
fakeroot alien --bump=0 --scripts --to-deb --target=$debarch \
    $pkg1 $pkg2 $pkg3 $pkg4 $pkg5 $pkg6 $pkg7 \
    $pkg8 $pkg9 $pkg10 || exit 1; \
rm -f ${path_prepend}/dh_shlibdeps; \
rmdir ${path_prepend}; \
rm -f $pkg1 $pkg2 $pkg3 $pkg4 $pkg5 $pkg6 $pkg7 \
    $pkg8 $pkg9 $pkg10;
Package build failed. Here's the log:
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
make[1]: Entering directory '/home/avery/build/kernels/z2.0.4-k5.11.6/zfs-2.0.4-k5.11.6/zfs-2.0.4'
dh 
dh: error: specify a sequence to run
make[1]: *** [debian/rules:7: binary] Error 25
make[1]: Leaving directory '/home/avery/build/kernels/z2.0.4-k5.11.6/zfs-2.0.4-k5.11.6/zfs-2.0.4'
make: *** [Makefile:1706: deb-utils] Error 1
``` 

Here's the package versions:

```
$ apt-show-versions | egrep 'alien|debhelper'

alien:all/bullseye 8.95.3 uptodate
debhelper:all/bullseye 13.3.4 uptodate
libdebhelper-perl:all/bullseye 13.3.4 uptodate
```

> Patching debian/rules in an _emitted_ debian/ directory from alien lets the package build successfully.

I don't know this terminology, what is an emitted directory?  Where might I find this location to patch?

It didn't help my problem with Debian that I copied the directories to an Ubuntu machine and finished compiling there, but at least it got done.

--- a/config/deb.am
+++ b/config/deb.am
@@ -15,13 +15,18 @@ deb-local:
 	"*** re-run configure, and try again.\n"; \
 		exit 1; \
 	fi; \
-	if test "${ALIEN_VERSION}" = "8.95.3"; then \
-		echo -e "\n" \
-	"*** Installed version of ${ALIEN} is known to be broken;\n" \
-	"*** attempting to generate debs will fail! See\n" \
-	"*** https://github.com/openzfs/zfs/issues/11650 for details.\n"; \
-		exit 1; \
-	fi)
+        if test "${ALIEN_MAJOR}" = "8" && \
+           test "${ALIEN_MINOR}" = "95"; then \
+        if test "${ALIEN_POINT}" = "1" || \
+           test "${ALIEN_POINT}" = "2" || \
+           test "${ALIEN_POINT}" = "3"; then \
+                /bin/echo -e "\n" \
+        "*** Installed version of ${ALIEN} is known to be broken;\n" \
+        "*** attempting to generate debs will fail! See\n" \
+        "*** https://github.com/openzfs/zfs/issues/11650 for details.\n"; \
+                exit 1; \
+        fi; \
+        fi)
 
 deb-kmod: deb-local rpm-kmod
 	name=${PACKAGE}; \

--- a/config/deb.am
+++ b/config/deb.am
@@ -14,6 +14,13 @@ deb-local:
 	"*** package for your distribution which provides ${ALIEN},\n" \
 	"*** re-run configure, and try again.\n"; \
 		exit 1; \
+	fi; \
+	if test "${ALIEN_VERSION}" = "8.95.3"; then \
+		echo -e "\n" \
+	"*** Installed version of ${ALIEN} is known to be broken;\n" \
+	"*** attempting to generate debs will fail! See\n" \
+	"*** https://github.com/openzfs/zfs/issues/11650 for details.\n"; \
+		exit 1; \
 	fi)
 
 deb-kmod: deb-local rpm-kmod

--- a/config/zfs-build.m4
+++ b/config/zfs-build.m4
@@ -437,6 +437,9 @@ AC_DEFUN([ZFS_AC_ALIEN], [
 	AC_MSG_CHECKING([whether $ALIEN is available])
 	AS_IF([tmp=$($ALIEN --version 2>/dev/null)], [
 		ALIEN_VERSION=$(echo $tmp | $AWK '{ print $[3] }')
+		ALIEN_MAJOR=$(echo ${ALIEN_VERSION} | $AWK -F'.' '{ print $[1] }')
+		ALIEN_MINOR=$(echo ${ALIEN_VERSION} | $AWK -F'.' '{ print $[2] }')
+		ALIEN_POINT=$(echo ${ALIEN_VERSION} | $AWK -F'.' '{ print $[3] }')
 		HAVE_ALIEN=yes
 		AC_MSG_RESULT([$HAVE_ALIEN ($ALIEN_VERSION)])
 	],[
@@ -447,6 +450,9 @@ AC_DEFUN([ZFS_AC_ALIEN], [
 	AC_SUBST(HAVE_ALIEN)
 	AC_SUBST(ALIEN)
 	AC_SUBST(ALIEN_VERSION)
+	AC_SUBST(ALIEN_MAJOR)
+	AC_SUBST(ALIEN_MINOR)
+	AC_SUBST(ALIEN_POINT)
 ])
 
 dnl #

